spec.bs

```diff
@@ -0,0 +1,393 @@
+</pre>
+
+Introduction {#intro}
+=====================
+
+<em>This section is non-normative.</em>
+
+Motivation {#motivation}
+------------------------
+
+Browsers are now working to prevent cross-site user tracking, including by
+partitioning storage and removing third-party cookies. There are a range of APIs
+proposals to continue supporting legitimate use cases in a way that respects
+user privacy. Many of these APIs, including <a href="https://wicg.github.io/shared-storage/">Shared Storage</a> and <a href="https://wicg.github.io/turtledove/">FLEDGE</a>, isolate
+potentially identifying cross-site data in special contexts, which ensures that
+the data cannot escape the user agent.
+
+Relative to cross-site data from an individual user, aggregate data about groups
+of users can be less sensitive and yet would be sufficient for a wide range of
+use cases. An aggregation service has been proposed to allow reporting noisy,
+aggregated cross-site data. This service was originally proposed for use by the
+Attribution Reporting API, but allowing more general aggregation would support
+additional use cases. In particular, the FLEDGE and Shared Storage proposals
+expect this functionality to become available.
+
+Overview {#overview}
+--------------------
+
+This document oulines a general-purpose API that can be called from isolated
+contexts that have access to cross-site data (such as a Shared Storage worklet).
+Within these contexts, potentially identifying data can be encapsulated into
+"aggregatable reports". To prevent leakage, the cross-site data in these reports
+is encrypted to ensure it can only be processed by the aggregation service.
+During processing, this service adds noise and imposes limits on how many
+queries can be performed.
+
+This API provides functions allowing the origin to construct an aggregatable
+report and specify the values to be embedded into its encrypted payload (for
+later computation via the aggregation service). These calls result in the
+aggregatable report being queued to be sent to the reporting endpoint of the
+script's origin after a delay. After the endpoint receives the reports, it will
+batch the reports and send them to the aggregation service for processing. The
+output of that process is a summary report containing the (approximate) result,
+which is dispatched back to the script's origin.
+
+General methods {#general-methods}
+==================================
+
+Worklet interface {#worklet-interface}
+--------------------------------------
+
+<xmp class="idl">
+[Exposed=(FledgeWorkletGlobalScope,SharedStorageWorkletGlobalScope)]
+interface PrivateAggregation {
+  undefined sendHistogramReport(PAHistogramContribution contribution);
+};
+
+[Exposed=(FledgeWorkletGlobalScope,SharedStorageWorkletGlobalScope)]
+dictionary PAHistogramContribution {
+  required bigint bucket;
+  required long value;
+};
+</xmp>
+
+Each {{PrivateAggregation}} has a <dfn>contributions cache</dfn>, a [=list=].
```

Comment:
Can we be a little more explicit, like a list of PAHistogramContributions?

Reply:
Alternatively we can define a type which is either PAHistogramContribution or PAExtendedHistogramContribution to avoid duplication on maintaining two caches. Wdyt?

Reply:
Added some prose to this effect